dali/pipeline/data/tensor_vector.cc

```diff
@@ -431,285 +666,354 @@ bool TensorVector<Backend>::is_pinned() const {
 
 
 template <typename Backend>
-int TensorVector<Backend>::device_id() const {
-  if (IsContiguous()) {
-    return tl_->device_id();
-  } else if (!tensors_.empty()) {
-    return tensors_[0]->device_id();
+void TensorVector<Backend>::set_device_id(int device_id) {
+  contiguous_buffer_.set_device_id(device_id);
+  for (auto &t : tensors_) {
+    t.set_device_id(device_id);
   }
-  return CPU_ONLY_DEVICE_ID;
+  device_ = device_id;
+}
+
+
+template <typename Backend>
+int TensorVector<Backend>::device_id() const {
+  return device_;
 }
 
 
 template <typename Backend>
 void TensorVector<Backend>::reserve(size_t total_bytes) {
-  if (state_ == State::noncontiguous) {
+  int batch_size_bkp = curr_num_tensors_;
+  if (!state_.IsContiguous()) {
     tensors_.clear();
-    curr_num_tensors_ = 0;
+    resize_tensors(0);
+  }
+  state_.Setup(BatchState::Contiguous);
+  contiguous_buffer_.reserve(total_bytes);
+  if (IsValidType(type_)) {
+    resize_tensors(batch_size_bkp);
+    recreate_views();
   }
-  state_ = State::contiguous;
-  tl_->reserve(total_bytes);
-  UpdateViews();
 }
 
 
 template <typename Backend>
 void TensorVector<Backend>::reserve(size_t bytes_per_sample, int batch_size) {
   assert(batch_size > 0);
-  state_ = State::noncontiguous;
+  state_.Setup(BatchState::Noncontiguous);
   resize_tensors(batch_size);
```

Comment:
Technically, a "reserve" function can specify smaller batch size than current one - in which case we should not shrink...